server/webapp/WEB-INF/rails/app/controllers/api/web_hooks/hosted_bit_bucket_controller.rb

```diff
@@ -0,0 +1,109 @@
+##########################################################################
+# Copyright 2017 ThoughtWorks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+##########################################################################
+
+module Api
+  module WebHooks
+    class HostedBitBucketController < WebHookController
+      before_action :prempt_ping_call
+      before_action :verify_content_origin
+      before_action :allow_only_push_event
+      before_action :allow_only_git_scm
+      before_action :verify_payload
+
+      protected
+
+      def repo_branch
+        payload['changes'].find {|change| change['ref']['type'] == 'BRANCH'}['ref']['displayId']
+      rescue
+        nil
+      end
+
+      def possible_urls
+        payload['repository']['links']['clone'].collect {|l| without_credentials(l['href'])}
```

Comment:
I think it is better to follow the contract instead of overriding `possible_urls`. Please consider implementing `repo_host_name`, `repo_full_name` and `repo_branch` over `possible_urls`.

I already went through the PR description about hosted version sending clone URLs with credentials. Please can you include he `Webhook` payload json` post reducting sensitive information.

Reply:
I'll check again if I can fit it into the existing contract - on my first attempt to write this, it didn't fit very smoothly with using a list of possible urls to construct the final urls. But I recognize that the current approach of stuffing it into the contract is not very robust.

Regarding a sample payload: the one in the `should call the material update service upon receiving a good request and respond with 202 [accepted]` test is exactly such a redacted payload. If that's not sufficient, please let me know.

Reply:
Checked the details again, and the main issue is that our bitbucket server returns _varying_ urls, so I cannot extract a _single_ `repo_host_name` that fits. To give you an example, this is what it returns in the clone URL list:
```
"clone": [
  {
    "href": "https://sso.my-company.com/bitbucket/scm/my-project/my-repo.git",
    "name": "http"
  },
  {
    "href": "ssh://git@git.my-company.com:7999/my-project/my-repo.git",
    "name": "ssh"
  }
],
```

So the full host domain differs:
- sso.my-company.com
- git.my-company.com:7999

I can cover that by extracting just the top domain (my-company.com) and adding specific templates to the `possible_urls` list, but that would be way too specific to my company's use case, in my opinion.

So I'm inclined to adapt the contract/hierarchy to support hooks that send a list of ready-made clone urls.
Something like:
```
WebHookController
  \- HostedBitBucketController
  \- GuessUrlWebHookController
       \- GitHubController
       \- BitBucketController
       \- GitLabController
```

Where `WebHookController` defers `possible_urls`, `repo_branch`, and `repo_name` (the latter for logging only) to its subclasses,
and `GuessUrlWebHookController` implements `possible_urls` as before with a list of possible urls and defers the necessary fields in the url templates to its subclasses.

Alternatively we could adapt just the resolution of the url templates in `possible_urls`, to allow for a _list_ returned in `repo_host_name`, which then generates a potentially very large cross-product of all host name / url combinations.

What do you think?

Reply:
> ```
> WebHookController
> \- HostedBitBucketController
>  \- GuessUrlWebHookController
>       \- GitHubController
>       \- BitBucketController
>       \- GitLabController
>```

@sandro-h - Sorry for the late response. I think it is a good idea to follow the approach that you mentioned above.

Reply:
I've pushed the discussed changes. I can squash the commits afterwards, if you wish.

> Where WebHookController defers possible_urls, repo_branch, and repo_name (the latter for logging only) to its subclasses,

I renamed the intended `repo_name` method to `repo_log_name`, to more clearly mark its purpose because it cannot be used for any real logic besides logging (since its content is somewhat arbitrary depending on the webhook type).

One small annoyance for the hosted Bitbucket webhook is that there's no robust way to extract a `repo_log_name` containing the project (e.g. my-project). Hosted Bitbucket only returns the repo name in a separate field and parsing the URL for the project name doesn't sound very robust.
Therefore the `repo_log_name` for the HostedBitBucket hook only contains a host name and repo name.

Reply:
@sandro-h - I will take a look at your changes next week.